### PR TITLE
Add create group operator

### PR DESCRIPTION
**Pull request by glopesdev**
This PR introduces a new `CreateGroup` operator used to convert the values in a sequence into a grouped observable, using each element of the input sequence as the group key. The values of each created group are specified through a nested workflow which receives the key value as its input.

### CODE REVIEW

**Review by aspaNeuro — APPROVED** [review]
Are you thinking in using this group operator for the modular implementation of the devices?
If so I would like to understand better how. And what's the difference between this operator and the group by.

**Comment by glopesdev** [comment]
@aspaNeuro yes that is also the goal. The main difference with `GroupBy` is that you can create a sequence of grouped observables directly from the key, instead of having to do `PairValueWithKey > GroupBy`. Graphically below it is the difference between the left-hand and right-hand side workflows:

image

Notice that in the earlier workaround we were abusing `GroupBy` to construct a grouped observable by having effectively sequences where all items are the same key. There were costs associated with this, e.g. we are creating a dictionary with a single entry for each sequence and then doing a lookup on that dictionary for every single value. Granted, these are mostly negligible but more importantly we want to be able to specify for every group what is the nested workflow that generates values, which requires some kind of nested operator that captures this pattern, which is hard to do simply using the existing operators.

`CreateGroup` also works for changing the values of existing groups while keeping the same key, which is also currently clunky to do. Graphically again the difference between the left-hand and right-hand side workflows:

image

Basically in case the input is an existing grouped observable, the key is taken from the group key and the elements are simply the sequence defined in the nested workflow.

**Comment by aspaNeuro** [comment]
Great, really neat and it seems quite easy to work with.


__

André Almeida

Tel: +44 (0) 77 3736 2319

Email: ***@***.******@***.***>

Website: https://neurogears.org<https://neurogears.org/>

Address: 31 Oval Road, London, NW1 7EA



NeuroGEARS Ltd is a limited company registered in England and Wales.

Company Registration No. 10773335

________________________________

**Comment by PathogenDavid** [comment]
This PR was merged, but is not actually present in `main`. [The merge commit](https://github.com/SainsburyWellcomeCentre/aeon_acquisition/commit/184ae27a19e0f2c489bac083446f62469988c703) is not part of this repository anymore, so I think `main` got force pushed at some point.
